### PR TITLE
refactor(source): Prepare for new PackageIDSpec syntax

### PR DESCRIPTION
**Pull request by epage**
### What does this PR try to resolve?

This adds tests and refactors to prepare for #12933

### CODE REVIEW

File: src/cargo/core/source_id.rs [hunk 1]
```diff
@@ -752,6 +752,20 @@ impl PartialEq for SourceIdInner {
     }
 }
 
+impl SourceKind {
+    pub fn protocol(&self) -> Option<&str> {
```
Comment (weihanglo) [inline src/cargo/core/source_id.rs:746]: Do we need to `pub` this?

It look like a implementation details for `Display` impl, especially that `SparseRegistry` return nothing here.

Comment (epage) [reply src/cargo/core/source_id.rs]: Huh, not sure why I made it `pub`.  Switched to `pub(crate)`.

**Review by weihanglo — APPROVED** [review]
All refactors look great now! Thank you :)

**Pull request merged**